sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -308,54 +300,93 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
         }
     }
 
-    @VisibleForTesting
-    protected IndexItemResponse indexItem(ShardUpsertRequest request,
-                                          ShardUpsertRequest.Item item,
-                                          IndexShard indexShard,
-                                          boolean tryInsertFirst,
-                                          UpdateSourceGen updateSourceGen,
-                                          InsertSourceGen insertSourceGen,
-                                          @Nullable ReturnValueGen returnGen,
-                                          boolean isRetry) throws Exception {
-        final long seqNo;
-        final long primaryTerm;
-        final long version;
-        Doc updatedDoc = null;
-        if (tryInsertFirst) {
-            version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE
-                ? Versions.MATCH_ANY
-                : Versions.MATCH_DELETED;
-            seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
-            primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
-            try {
-                item.source(insertSourceGen.generateSourceAndCheckConstraints(item.insertValues()));
-            } catch (IOException e) {
-                throw ExceptionsHelper.convertToElastic(e);
-            }
-        } else {
-            Doc currentDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
-
-            Map<String, Object> updatedSource = updateSourceGen.generateSource(
-                currentDoc,
-                item.updateAssignments(),
-                item.insertValues()
+    protected IndexItemResponse insert(@Nullable ReturnValueGen returnGen,
+                                     @Nullable InsertSourceGen insertSourceGen,
+                        ShardUpsertRequest request,
+                        ShardUpsertRequest.Item item,
+                        IndexShard indexShard,
+                        boolean isRetry) throws Exception {
+        assert insertSourceGen != null : "InsertSourceGen must not null";
+        long version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE ? Versions.MATCH_ANY : Versions.MATCH_DELETED;
+        long seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+        long primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+        BytesReference rawSource;
+        try {
+            rawSource = insertSourceGen.generateSourceAndCheckConstraints(item.insertValues());
+            item.source(rawSource);
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+        Engine.IndexResult indexResult = index(seqNo, primaryTerm, version, isRetry, indexShard, item);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            Map<String, Object> source = JsonXContent.jsonXContent.createParser(
```

Comment:
This is the case where the source needs to be converted from raw to Map<String, Object> if we want evaluate return values.